system-addon/content-src/components/TopSites/TopSites.jsx

```diff
@@ -69,6 +70,49 @@ class TopSite extends React.Component {
   }
 }
 
+/**
+ * A proxy class that uses double requestAnimationFrame from
+ * componentDidMount to send an async message to parent with performance
+ * data after the paint/
+ *
+ * XXX If TopSites can be changed to extend React.PureComponent, both that
+ * and this should be too.
+ * XXX Should probably be split out into a higher-order component to wrap
+ * whatever.
+ *
+ * @class TopSitesPerfTimer
+ * @extends {React.Component}
+ */
+class TopSitesPerfTimer extends React.Component {
+  constructor(props) {
+    super(props);
+    this.perfSvc = this.props.perfSvc || perfSvc;
+    this.sendAsyncMessage = this.props.sendAsyncMessage || window.sendAsyncMessage;
+  }
+  componentDidMount() {
+    this._onNextFrame(this._sendPaintedEvent.bind(this));
+  }
+  _onNextFrame(callback) {
+    requestAnimationFrame(() => {
+      requestAnimationFrame(callback);
+    });
+  }
+  _sendPaintedEvent() {
+    this.perfSvc.mark("topsites_painted_ts");
+
+    let topsites_painted_ts = this.perfSvc
+      .getMostRecentAbsMarkStartByName("topsites_painted_ts");
+
+    this.sendAsyncMessage("ActivityStream:ContentToMain", {
```

Comment:
In this context, we should be able to use dispatch the action without needing "ContentToMain" stuff.